File: apps/backend-gateway/src/application/services/stadium_service.py
"""
StadiumOS AI — Stadium & Sector Services.

Coordinates use cases involving stadium profiles and sector boundaries.
"""

import uuid
import structlog

from src.application.ports.stadium_repository import StadiumRepository
from src.application.schemas.stadium_schemas import (
    SectorCreate,
    SectorResponse,
    StadiumCreate,
    StadiumResponse,
)
from src.domain.entities.sector import Sector
from src.domain.entities.stadium import Stadium
from src.domain.value_objects import CapacityInfo, Coordinates
from src.domain.exceptions import DuplicateEntityError, EntityNotFoundError

logger = structlog.get_logger("stadium_service")


class StadiumService:
    """
    Coordinates CRUD operations for Stadiums and Sectors.
    """

    def __init__(self, repo: StadiumRepository) -> None:
        self.repo = repo

    async def create_stadium(self, req: StadiumCreate) -> StadiumResponse:
        existing = await self.repo.get_by_name(req.name)
        if existing:
            raise DuplicateEntityError("Stadium", "name", req.name)

        new_stadium = Stadium(
            id=uuid.uuid4(),
            name=req.name,
            city=req.city,
            country=req.country,
            location=Coordinates(latitude=req.latitude, longitude=req.longitude),
            total_capacity=req.total_capacity,
            timezone=req.timezone,
        )

        saved = await self.repo.save(new_stadium)
        await logger.ainfo("stadium_created", stadium_id=str(saved.id), name=saved.name)
        
        # Return mapper model validation
        return StadiumResponse(
            id=saved.id,
            name=saved.name,
            city=saved.city,
            country=saved.country,
            latitude=saved.location.latitude,
            longitude=saved.location.longitude,
            total_capacity=saved.total_capacity,
            timezone=saved.timezone,
        )

    async def get_stadium(self, stadium_id: uuid.UUID) -> StadiumResponse:
        stadium = await self.repo.get_by_id(stadium_id)
        if not stadium:
            raise EntityNotFoundError("Stadium", str(stadium_id))

        return StadiumResponse(
            id=stadium.id,
            name=stadium.name,
            city=stadium.city,
            country=stadium.country,
            latitude=stadium.location.latitude,
            longitude=stadium.location.longitude,
            total_capacity=stadium.total_capacity,
            timezone=stadium.timezone,
        )

    async def list_stadiums(self) -> list[StadiumResponse]:
        stadiums = await self.repo.list_all()
        return [
            StadiumResponse(
                id=s.id,
                name=s.name,
                city=s.city,
                country=s.country,
                latitude=s.location.latitude,
                longitude=s.location.longitude,
                total_capacity=s.total_capacity,
                timezone=s.timezone,
            )
            for s in stadiums
        ]

    async def delete_stadium(self, stadium_id: uuid.UUID) -> None:
        stadium = await self.repo.get_by_id(stadium_id)
        if not stadium:
            raise EntityNotFoundError("Stadium", str(stadium_id))

        await self.repo.delete(stadium_id)
        await logger.ainfo("stadium_deleted", stadium_id=str(stadium_id))

    # --- Sector Actions ---

    async def create_sector(self, stadium_id: uuid.UUID, req: SectorCreate) -> SectorResponse:
        # Verify stadium exists
        stadium = await self.repo.get_by_id(stadium_id)
        if not stadium:
            raise EntityNotFoundError("Stadium", str(stadium_id))

        # Verify duplicate sector name in this stadium
        sectors = await self.repo.list_sectors_by_stadium(stadium_id)
        if any(sec.name.lower() == req.name.lower() for sec in sectors):
            raise DuplicateEntityError("Sector", "name", req.name)

        new_sector = Sector(
            id=uuid.uuid4(),
            stadium_id=stadium_id,
            name=req.name,
            capacity_info=CapacityInfo(
                max_capacity=req.max_capacity,
                warning_threshold_percent=req.warning_threshold_percent,
                critical_threshold_percent=req.critical_threshold_percent,
            ),
            centroid=Coordinates(latitude=req.centroid_latitude, longitude=req.centroid_longitude),
            is_accessible=req.is_accessible,
        )

        saved = await self.repo.save_sector(new_sector)
        await logger.ainfo("sector_created", sector_id=str(saved.id), name=saved.name, stadium_id=str(stadium_id))

        return SectorResponse(
            id=saved.id,
            stadium_id=saved.stadium_id,
            name=saved.name,
            max_capacity=saved.capacity_info.max_capacity,
            warning_threshold_percent=saved.capacity_info.warning_threshold_percent,
            critical_threshold_percent=saved.capacity_info.critical_threshold_percent,
            centroid_latitude=saved.centroid.latitude,
            centroid_longitude=saved.centroid.longitude,
            is_accessible=saved.is_accessible,
        )

    async def list_sectors(self, stadium_id: uuid.UUID) -> list[SectorResponse]:
        stadium = await self.repo.get_by_id(stadium_id)
        if not stadium:
            raise EntityNotFoundError("Stadium", str(stadium_id))

        sectors = await self.repo.list_sectors_by_stadium(stadium_id)
        return [
            SectorResponse(
                id=s.id,
                stadium_id=s.stadium_id,
                name=s.name,
                max_capacity=s.capacity_info.max_capacity,
                warning_threshold_percent=s.capacity_info.warning_threshold_percent,
                critical_threshold_percent=s.capacity_info.critical_threshold_percent,
                centroid_latitude=s.centroid.latitude,
                centroid_longitude=s.centroid.longitude,
                is_accessible=s.is_accessible,
            )
            for s in sectors
        ]

    async def delete_sector(self, sector_id: uuid.UUID) -> None:
        sector = await self.repo.get_sector_by_id(sector_id)
        if not sector:
            raise EntityNotFoundError("Sector", str(sector_id))

        await self.repo.delete_sector(sector_id)
        await logger.ainfo("sector_deleted", sector_id=str(sector_id))

File: apps/backend-gateway/src/application/services/stadiumService.ts
/**
 * StadiumOS AI — Stadium & Sector Management Service.
 * 
 * Orchestrates creating, reading, and verifying stadium venues and sectors.
 */

import { v4 as uuidv4 } from "uuid";
import { IStadiumRepository } from "../ports/stadiumRepository";
import { StadiumCreate, SectorCreate } from "../schemas/stadiumSchemas";
import { Stadium } from "../../domain/entities/stadium";
import { Sector } from "../../domain/entities/sector";
import { Coordinates, CapacityInfo } from "../../domain/valueObjects";
import { DuplicateEntityError, EntityNotFoundError } from "../../domain/exceptions";
import { logger } from "../../infrastructure/logging/logger";

export class StadiumService {
  private stadiumRepo: IStadiumRepository;

  constructor(stadiumRepo: IStadiumRepository) {
    this.stadiumRepo = stadiumRepo;
  }

  public async createStadium(req: StadiumCreate): Promise<Stadium> {
    const existing = await this.stadiumRepo.getByName(req.name);
    if (existing) {
      throw new DuplicateEntityError("Stadium", "name", req.name);
    }

    const newStadium = new Stadium(
      uuidv4(),
      req.name,
      req.city,
      req.country,
      new Coordinates(req.latitude, req.longitude),
      req.total_capacity,
      req.timezone
    );

    const saved = await this.stadiumRepo.save(newStadium);
    logger.info({ stadiumId: saved.id, name: saved.name }, "Stadium venue created successfully");
    return saved;
  }

  public async getStadium(id: string): Promise<Stadium> {
    const stadium = await this.stadiumRepo.getById(id);
    if (!stadium) {
      throw new EntityNotFoundError("Stadium", id);
    }
    return stadium;
  }

  public async listStadiums(): Promise<Stadium[]> {
    return await this.stadiumRepo.listAll();
  }

  public async createSector(stadiumId: string, req: SectorCreate): Promise<Sector> {
    // Assert stadium exists
    await this.getStadium(stadiumId);

    const newSector = new Sector(
      uuidv4(),
      stadiumId,
      req.name,
      new CapacityInfo(req.max_capacity, req.warning_threshold_percent, req.critical_threshold_percent),
      new Coordinates(req.centroid_latitude, req.centroid_longitude),
      req.is_accessible
    );

    const saved = await this.stadiumRepo.saveSector(newSector);
    logger.info({ sectorId: saved.id, name: saved.name }, "Stadium sector created successfully");
    return saved;
  }

  public async getSector(id: string): Promise<Sector> {
    const sector = await this.stadiumRepo.getSectorById(id);
    if (!sector) {
      throw new EntityNotFoundError("Sector", id);
    }
    return sector;
  }

  public async listSectors(stadiumId: string): Promise<Sector[]> {
    await this.getStadium(stadiumId);
    return await this.stadiumRepo.listSectorsByStadium(stadiumId);
  }
}
export default StadiumService;
